### PR TITLE
ENH: implement mean and triangle thresholding

### PR DESCRIPTION
**Pull request by sciunto**
## Description

Replaces #2124 (was too difficult to rebase)

### CODE REVIEW

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -490,3 +490,111 @@ def find_local_maxima(hist):
             return bin_centers[upper_bound]
         elif bias == 'mid':
             return bin_centers[(threshold + upper_bound) // 2]
+
+
+def threshold_mean(image):
```
Comment (ahojnnes) [inline skimage/filters/thresholding.py]: Example?

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -490,3 +490,111 @@ def find_local_maxima(hist):
...
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_mean(image)
+    >>> binary = image > thresh
+    """
+    assert_nD(image, 2)
+    return np.mean(image)
+
+
+def threshold_triangle(image, nbins=256):
```
Comment (ahojnnes) [inline skimage/filters/thresholding.py]: Example in the gallery?

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (sciunto) [inline skimage/filters/thresholding.py]: @ahojnnes As said in the main discussion, I'll refactor the examples later.

Comment (soupault) [inline skimage/filters/thresholding.py]: @ahojnnes see https://github.com/scikit-image/scikit-image/pull/2124#issuecomment-223073043 and https://github.com/scikit-image/scikit-image/pull/2124#issuecomment-223081464 .

Comment (sciunto) [inline skimage/filters/thresholding.py]: I started to do the job in #2110

**Comment by ahojnnes** [comment]
LGTM.

File: skimage/filters/tests/test_thresholding.py [hunk 3]
```diff
@@ -311,5 +313,37 @@ def test_threshold_minimum_failure():
...
+    t_inv_img = inv_img > t
+    t_inv_inv_img = np.invert(t_inv_img)
+
+    t = threshold_triangle(img)
+    t_img = img > t
+
+    # Check that most of the pixels are identical
+    # See numpy #7685 for a future np.testing API
+    unequal_pos = np.where(t_img.ravel() != t_inv_inv_img.ravel())
+    assert(len(unequal_pos[0]) / len(t_img.ravel()) < 1e-2)
```
Comment (soupault) [inline skimage/filters/tests/test_thresholding.py]: What is the reasoning behind this value?
Also, `len(t_img.ravel())` == `t_img.size`.

File: skimage/filters/thresholding.py [hunk 4]
```diff
@@ -490,3 +490,111 @@ def find_local_maxima(hist):
...
+    # nbins is ignored for interger arrays
+    # so, we recalculate the effective nbins.
+    hist, bin_centers = histogram(image.ravel(), nbins)
+    nbins = bin_centers[-1] - bin_centers[0] + 1
+
+    # Find peak, lowest and highest gray levels.
+    arg_peak_height = np.argmax(hist)
+    peak_height = hist[arg_peak_height]
+    arg_low_level = np.where(hist>0)[0][0]
+    arg_high_level = np.where(hist>0)[0][-1]
```
Comment (soupault) [inline skimage/filters/thresholding.py]: I'd speed-up this a bit as follows:
`arg_low_level, arg_high_level = np.where(hist>0)[0][[0, -1]]`.
Up to you, anyways.

File: skimage/filters/thresholding.py [hunk 5]
```diff
@@ -490,3 +490,111 @@ def find_local_maxima(hist):
...
+        vol. 55, pp. 532-537, 1993.
+
+    Examples
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_mean(image)
+    >>> binary = image > thresh
+    """
+    assert_nD(image, 2)
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Why constrain this to 2D only? I think, any-D single-channel arrays should be supported.

Comment (sciunto) [inline skimage/filters/thresholding.py]: I realized I confused myself on the meaning of nD. I meant grayscale. We really need to document things around nD supports...

File: skimage/filters/tests/test_thresholding.py (hunk 3 above)
Comment (sciunto) [inline skimage/filters/tests/test_thresholding.py]: the value just sounds reasonable...

File: skimage/filters/thresholding.py (hunk 5 above)
Comment (soupault) [inline skimage/filters/thresholding.py]: Yeah, right after the implementation of fused types :smile:.

File: skimage/filters/thresholding.py [hunk 6]
```diff
@@ -490,3 +490,108 @@ def find_local_maxima(hist):
...
             return bin_centers[(threshold + upper_bound) // 2]
+
+
+def threshold_mean(image):
+    """Return threshold value based on the mean of grayscale values.
+
+    Parameters
+    ----------
+    image : (N, M) ndarray
+        Grayscale input image.
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Need to change the signature -> `(N, M[, ..., P])` and description.

File: skimage/filters/thresholding.py (hunk 5 above)
Comment (sciunto) [inline skimage/filters/thresholding.py]: WHat do you mean by fused types? Is  there an open ticket to know more?

Comment (soupault) [inline skimage/filters/thresholding.py]: Well, I think the solution to our regular issues could be to have a set of fused types like: ImageGrayscale2D, ImageGrayscale3D, ImageMultichannel3D, ImageGrayscaleND, ... or something more detailed to cover also never-ending issues with different dtype support and convertion: ImageFloat, ImageUint8, ...
The majority of our algorithms has limited support in terms of dimensionality and dtype (some support integer types, other support only floats). My personal opinion is that using this kind of a wrapper we could handle all of this cases more clearly. I'm aware if this discussion has already been carried out.
As an example, you might consider `pandas` with their `Dataframes` and fused dtypes for data columns, indices, etc (e.g. https://github.com/pydata/pandas/blob/master/pandas/types/dtypes.py).

Comment (soupault) [inline skimage/filters/thresholding.py]: We have a small piece of ambitions [here](http://scikit-image.org/docs/stable/user_guide/numpy_images.html#a-note-on-time).
Frankly speaking, 4D+ spatial data does not make sense to me at all. I suppose, the dream is to have 1timeD+3spatialD+1channelD for all image processing algorithms, but that is, of course, not really feasible. The main message is: if it is easy and cheap to add 3spatialD (or even NspatialD) support, one shouldn't hesitate to do that. :grey_exclamation:

**Comment by sciunto** [comment]
@soupault Fixed ;)

**Comment by soupault** [comment]
@sciunto please, squash commits and we're done.

**Comment by sciunto** [comment]
squashed :)

**Comment by soupault** [comment]
@sciunto Thanks! I've added a requirement to the checklist in https://github.com/scikit-image/scikit-image/pull/2110 .

**Pull request merged**